p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,350 @@
+		Negotiated: &negotiated,
+		ID:         &id,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	errC := awaitPeerConnectionOpen(addr.ufrag, pc)
+	// we infer the client sdp from the incoming STUN connectivity check
+	// by setting the ice-ufrag equal to the incoming check.
+	clientSdpString := internal.RenderClientSDP(addr.raddr, addr.ufrag)
+	clientSdp := webrtc.SessionDescription{SDP: clientSdpString, Type: webrtc.SDPTypeOffer}
+	pc.SetRemoteDescription(clientSdp)
+
+	answer, err := pc.CreateAnswer(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	err = pc.SetLocalDescription(answer)
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-errC:
+		if err != nil {
+			return nil, fmt.Errorf("peer connection error: %w", err)
+		}
+
+	}
+
+	rwc, err := internal.GetDetachedChannel(ctx, rawDatachannel)
+	if err != nil {
+		return nil, err
+	}
+
+	handshakeChannel := newStream(nil, rawDatachannel, rwc, l.localAddr, addr.raddr)
+	// The connection is instantiated before performing the Noise handshake. This is
+	// to handle the case where the remote is faster and attempts to initiate a stream
+	// before the ondatachannel callback can be set.
+	conn, err := newConnection(
+		network.DirInbound,
+		pc,
+		l.transport,
+		scope,
+		l.transport.localPeerId,
+		l.transport.privKey,
+		l.localMultiaddr,
+		"",  // remotePeer
+		nil, // remoteKey
+		remoteMultiaddr,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// we do not yet know A's peer ID so accept any inbound
+	secureConn, err := l.transport.noiseHandshake(ctx, pc, handshakeChannel, "", crypto.SHA256, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// earliest point where we know the remote's peerID
+	err = scope.SetPeer(secureConn.RemotePeer())
+	if err != nil {
+		return nil, err
+	}
+
+	conn.setRemotePeer(secureConn.RemotePeer())
+	conn.setRemotePublicKey(secureConn.RemotePublicKey())
+
+	return conn, err
+}
+
+func (l *listener) Accept() (tpt.CapableConn, error) {
+	select {
+	case <-l.ctx.Done():
+		return nil, os.ErrClosed
+	case conn := <-l.acceptQueue:
+		return conn, nil
+	}
+}
+
+func (l *listener) Close() error {
+	select {
+	case <-l.ctx.Done():
+	default:
+		l.cancel()
+		l.wg.Wait()
+	}
+	return nil
+}
+
+func (l *listener) Addr() net.Addr {
+	return l.localAddr
+}
+
+func (l *listener) Multiaddr() ma.Multiaddr {
+	return l.localMultiaddr
+}
+
+func awaitPeerConnectionOpen(ufrag string, pc *webrtc.PeerConnection) <-chan error {
+	errC := make(chan error)
+	var once sync.Once
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		switch state {
+		case webrtc.PeerConnectionStateConnected:
+			once.Do(func() { close(errC) })
+		case webrtc.PeerConnectionStateFailed:
+			once.Do(func() {
+				// this ensures that we don't block this routine if the
+				// listener goes away
+				select {
+				case errC <- fmt.Errorf("peerconnection failed: %s", ufrag):
+					close(errC)
+				default:
```

Comment:
Why is this needed? Would be easier to just create `errC` with a cap of 1, wouldn't it?